### DISCUSSION TITLE
Spec First API Development + Pydantic Validators

### DISCUSSION THREAD

**Discussion by dsinghvi** [pydantic/pydantic#4789] [Question]
👋 Pydantic friends! I am working on [Fern](https://github.com/fern-api/fern) which makes it easier to adopt a schema-first API design process. Similar to OpenAPI, you define your API in a spec, and then generate code for servers + clients. Here is an example FastAPI microservice using Fern: [spec](https://github.com/fern-api/venus/blob/35ae881f0a0a4d275b8112f2fdc21e1cc9434d36/fern/venus-api/definition/organization.yml#L33-L61), [generated pydantic models](https://github.com/fern-api/venus/tree/main/src/venus/generated/server/resources/organization/types), [generated fastapi server interface](https://github.com/fern-api/venus/blob/35ae881f0a0a4d275b8112f2fdc21e1cc9434d36/src/venus/generated/server/resources/organization/service.py#L22-L42).

Our users want to be able to add validators to the generated Pydantic models. The catch is that they cannot modify the auto-generated classes because manual code changes would be erased upon regeneration.

Options to solve this problem: 
- **Dynamically run callables in pydantic.validator (status quo)**  Today, users can register Callables with generated models which will be invoked inside of a Pydantic validator. For example, in [create_organization_request.py](https://github.com/fern-api/venus/blob/35ae881f0a0a4d275b8112f2fdc21e1cc9434d36/src/venus/generated/server/resources/organization/types/create_organization_request.py#L41-L48):
  - users register their callables by using an annotation like `@CreateOrganizationRequest.Validators.field("organization_id")` ([code](https://github.com/fern-api/venus/blob/35ae881f0a0a4d275b8112f2fdc21e1cc9434d36/src/venus/generated/server/resources/organization/types/create_organization_request.py#L24-L38))
  - callables are stored for each field ([code](https://github.com/fern-api/venus/blob/35ae881f0a0a4d275b8112f2fdc21e1cc9434d36/src/venus/generated/server/resources/organization/types/create_organization_request.py#L40-L48))
  - we autogenerate Pydantic validator methods that call the aforementioned callables ([code](https://github.com/fern-api/venus/blob/35ae881f0a0a4d275b8112f2fdc21e1cc9434d36/src/venus/generated/server/resources/organization/types/create_organization_request.py#L98-L116))
  *A problem with this approach is that it doesn't leverage many validation features (pre, item, etc.) and doing so would make the generated code even more verbose.* <br>
- **Dynamically add validators to generated models**: I've seen a couple discussions ([here](https://github.com/pydantic/pydantic/discussions/2873) and [here](https://github.com/pydantic/pydantic/issues/825)) about how to add validators dynamically to the generated Pydantic model, but it seems like there are edge cases where that wouldn't work such as inheritance, root validators. It would be great if users could attach validator functions to models in a first class way -- is that something that would be possible to build? 

Thoughts on other options?

**Comment by samuelcolvin** [comment]
Humm the short answer is, I'm not sure.

But the change to V2 is quite relevant here. If you haven't seen, the [blog post](https://pydantic-docs.helpmanual.io/blog/pydantic-v2/) is a good start. `main` of this repo has already adopted pydantic-core, so might be worth look at how `main` works, it's a complete change from pydantic 1.10 and before.

It might be possible to provide a hook to insert pre-defined validators into the schema when it's being generated.

**Reply by dsinghvi** [reply]
> It might be possible to provide a hook to insert pre-defined validators into the schema when it's being generated.

What would this look like in Pydantic V2?

**Reply by samuelcolvin** [reply]
It would probably be a dunder method on the a model that, if defined would be called after creating the pydantic-core schema, and before creating the validator, you could implement it to insert functional validators into the nested schema.

The other option would be some way to customise [`ValidationFunctions`](https://github.com/pydantic/pydantic/blob/1ac1199f08c42803dc7c5e6fae754afd53e3f1b4/pydantic/_internal/_validation_functions.py#L42) to get validators from another source before creating the pydantic-core schema.

**Reply by samuelcolvin** [reply]
no reason why we couldn't implement both to provide more flexiblity.

**Reply by dsinghvi** [reply]
Got it -- so sounds like the `dunder method` would need to know about the user's custom validation methods. Today, the auto-generated model has no knowledge about validation methods that are going to be implemented by the user.

**Reply by dsinghvi** [reply]
But presumably we can dynamically add references to methods in the `dunder method` which could work

**Reply by samuelcolvin** [reply]
Not exactly.

The core validator (instance of `SchemaValidator`) cannot be altered. The only way to change the validation logic of a pydantic model in v2 will be to modify the pydantic-core schema, then call 

https://github.com/pydantic/pydantic/blob/1ac1199f08c42803dc7c5e6fae754afd53e3f1b4/pydantic/main.py#L454

It's probably worth having a look through the code on `main` now as that's already adopted pydantic-core for validation.

**Reply by dsinghvi** [reply]
@samuelcolvin sorry for the delayed response. i will dig through the code on main like you suggested to see if the dunder function or customization for `ValidationFunctions` makes more sense.
